pkg/analyzer/analyzers/airtable/airtablepat/airtable.go

```
@@ -0,0 +1,205 @@
+
+	if resp.StatusCode == http.StatusOK {
+		scopeStatusMap[scopeString] = true
+		return true, nil
+	} else if endpoint.ExpectedErrorResponse != nil {
+		var result map[string]interface{}
+		if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
+			return false, err
+		}
+
+		if errorInfo, ok := result["error"].(map[string]interface{}); ok {
+			if errorType, ok := errorInfo["type"].(string); ok && errorType == endpoint.ExpectedErrorResponse.Type {
+				scopeStatusMap[scopeString] = false
+				return false, nil
+			}
+		}
+	}
+
+	scopeStatusMap[scopeString] = true
+	return true, nil
+}
+
+func determineScopes(token string, basesInfo *common.AirtableBases) error {
+	if basesInfo != nil && len(basesInfo.Bases) > 0 {
+		for _, base := range basesInfo.Bases {
+			if base.Schema != nil && len(base.Schema.Tables) > 0 {
+				ids := map[string]string{"baseID": base.ID}
+				tableScopesDetermined := false
+
+				// Verify token "webhooks:manage" permission
+				_, err := determineScope(token, common.WebhookManage, ids)
+				if err != nil {
+					return err
+				}
+				// Verify token "block:manage" permission
+				_, err = determineScope(token, common.BlockManage, ids)
+				if err != nil {
+					return err
+				}
+
+				// Verifying scopes that require an existing table
+				for _, table := range base.Schema.Tables {
+					ids["tableID"] = table.ID
+
+					if !tableScopesDetermined {
+						_, err = determineScope(token, common.SchemaBasesWrite, ids)
+						if err != nil {
+							return err
+						}
+						_, err = determineScope(token, common.DataRecordsWrite, ids)
+						if err != nil {
+							return err
+						}
+						tableScopesDetermined = true
+					}
+
+					if granted, err := determineScope(token, common.DataRecordsRead, ids); err != nil {
+						return err
+					} else if granted {
+						// Verifying scopes that require an existing record and record read permission
+						records, err := fetchAirtableRecords(token, base.ID, table.ID)
```

question/suggestion: It looks like we might be again swallowing the error? If this is also done intentionally let's add a comment explaining why. We can make this a little easier to reason about if we check the err and handle it first. We then don't even need to check the `len(records)` we can simply range over them.

Ex:
```
records, err := fetchAirtableRecords(token, base.ID, table.ID)
if err != nil {
    return err // or continue
}
for _, record := range records {
    ids["recordID"] = record.ID
    _, err = determineScope(token, common.DataRecordcommentsRead, ids)
    if err != nil {
	    return err
    }
    break
}
```

---

Agreed, and there is actually one more condition required here. To avoid unnecessarily iterating over another `table` in the range `base.Schema.Tables`, I'll add another check after ranging through the records to check if `len(records) != 0`, as that indicates we have iterated through at least 1 `record` and determined the relevant scopes, therefore we'll `break` both the inner loop and outer loop.
So:
```
records, err := fetchAirtableRecords(token, base.ID, table.ID)
if err != nil {
    return err // or continue
}
for _, record := range records {
    ids["recordID"] = record.ID
    _, err = determineScope(token, common.DataRecordcommentsRead, ids)
    if err != nil {
	    return err
    }
    break
}
if len(records) != 0 {
    break
}
```